cookbooks/creating_a_behat_extension.rst

```diff
@@ -0,0 +1,270 @@
+Creating a Behat extension
+==========================
+
+Extensions are particularly useful when configuration becomes a necessity.
+
+In this cookbook, we will create a simple extension named ``LogsExtension`` that logs scenario durations and cover:
+
+#. Setting up the context
+#. Creating the extension
+#. Initializing the context
+
+Setting Up the Context
+----------------------
+
+First, we need to create a context that will handle the logging.
+The ``LogsExtension`` will provide a ``LogsContext`` that hooks into scenarios to log their start and end times.
+
+Directory structure:
+
+.. code-block::
+  
+  src/
+      Context/
+          LogsContext.php   # This is where we'll implement our logging logic
+
+
+The code for ``LogsContext.php``:
+
+.. code-block:: php
+
+  <?php
+  
+  namespace Behat\LogsExtension\Context;
+  
+  use Behat\Behat\Context\Context;
+  use Behat\Behat\Hook\Scope\BeforeScenarioScope;
+  use Behat\Behat\Hook\Scope\AfterScenarioScope;
+  
+  class LogsContext implements Context
+  {
+      /** @BeforeScenario */
+      public function before(BeforeScenarioScope $scope)
+      {
+          if (/* enable config */ === false) {
+              return;
+          }
+
+          file_put_contents(
+              /* filepath */,
+              'START: ' . $scope->getScenario()->getTitle() . ' - ' . time() . PHP_EOL,
+              FILE_APPEND
+          );
+      }
+  
+      /** @AfterScenario */
+      public function after(AfterScenarioScope $scope)
+      {
+          if (/* enable config */ === false) {
+              return;
+          }
+  
+          file_put_contents(
+              /* filepath */,
+              'END: ' . $scope->getScenario()->getTitle() . ' - ' . time() . PHP_EOL,
+              FILE_APPEND
+          );
+      }
+  }
+
+Creating the Extension
+----------------------
+
+Next, we need to create the extension itself.
+This will serve as the entry point for our logging functionality.
+
+Directory structure:
+
+.. code-block::
+  
+  src/
+      Context/
+          LogsContext.php
+      ServiceContainer/
+          LogsExtension.php   # This is where we'll define our extension
+
+To ensure Behat can find and load the ``LogsExtension.php`` file, it is important to place it within the `ServiceContainer` folder.
+While there might be alternatives, we will stick to the straightforward method.
```

Comment:
I'm not certain this is correct - I think it can be called & located anywhere that your composer autoloader can find it?

Reply:
This is the part where I was stuck, not understanding why my extension wouldn't load _(kudos @TimoBakx for helping me on Symfony's Slack 🙏)_.

From my research, extensions seems to be loaded via the [ExtensionManager](https://github.com/Behat/Behat/blob/42f15e5d513fc10dd807f90379652bacd56930b6/src/Behat/Testwork/ServiceContainer/ExtensionManager.php#L176-L198). `$locator` is the key value used to define the extension configuration in the `behat.yml` file.

There four ways for the manager to load the extension:
```php
        if (class_exists($class = $locator)) {
            return new $class;
        }

        if (class_exists($class = $this->getFullExtensionClass($locator))) {
            return new $class;
        }

        if (file_exists($locator)) {
            return require($locator);
        }

        if (file_exists($path = $this->extensionsPath . DIRECTORY_SEPARATOR . $locator)) {
            return require($path);
        }
```

But I didn't know why, it was not working on putting the full namespace of the class wasn't enough in my tests, I had to put it in the `ServiceContainer` folder.

So within the post I choose to go the easy way, but we can change it.
I can try to understand why it wasn't working at the first place.

Reply:
@monitaurus using the fully-qualified class for the extension does work and is applied before any of the subsequent magic in that method. If that wasn't working for you my best guess would be your class was in a file that composer couldn't autoload (so the `class_exists($locator)` at the top of that quoted code block would have returned false).

Reply:
I tried again, with the new `HelloWorld` extension, but it didn't work.
I'll try some step debug to find out, but I just let the full class namespace in the cookbook for simplicity.